### PR TITLE
adding removeContext

### PR DESCRIPTION
**Pull request by svejdo1**

### CODE REVIEW

**Comment by psiinon** [comment]
This file is generated by https://github.com/zaproxy/zaproxy/blob/develop/src/org/zaproxy/zap/extension/api/DotNetAPIGenerator.java, and so should not be manually updated.
Having said that, if its out of date we need to regenerate the API...

**Comment by svejdo1** [comment]
I'm not sure how to do that. However if you're going to do that somehow -  take a note that setting the api key is wrong  all over the code base:
	if (string.IsNullOrWhiteSpace(apikey)){
				parameters.Add("apikey", apikey);
	}

should actually by
	if (!string.IsNullOrWhiteSpace(apikey)){
				parameters.Add("apikey", apikey);
	}

**Comment by thc202** [comment]
That issue was addressed in zaproxy/zaproxy#2625, in theory the code just needs to be regenerated and released.

**Comment by kingthorin** [comment]
Please do an interactive rebase and fix this up into a single commit, https://github.com/zaproxy/zaproxy/blob/develop/CONTRIBUTING.md#what-we-zap-team-expect-from-you .

- https://www.kernel.org/pub/software/scm/git/docs/git-rebase.html#_interactive_mode
- https://wiki.eclipse.org/EGit/User_Guide#Interactive_Rebase

File: src/OWASPZAPDotNetAPI/.vs/config/applicationhost.config [hunk 1]
```diff
@@ -0,0 +1,1031 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (thc202) [inline src/OWASPZAPDotNetAPI/.vs/config/applicationhost.config]: Is this file really needed?

File: src/OWASPZAPDotNetAPI/OWASPZAPDotNetAPI/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -1,56 +1,19 @@
-﻿/* Zed Attack Proxy (ZAP) and its related class files.
```
Comment (thc202) [inline src/OWASPZAPDotNetAPI/OWASPZAPDotNetAPI/Properties/AssemblyInfo.cs]: Was this removed on purpose?

File: src/OWASPZAPDotNetAPI/OWASPZAPDotNetAPI/Properties/AssemblyInfo.cs [hunk 3]
```diff
@@ -1,56 +1,19 @@
...
+// General Information about an assembly is controlled through the following
 // set of attributes. Change these attribute values to modify the information
 // associated with an assembly.
-[assembly: AssemblyTitle("OWASPZAPDotNetAPI")]
-[assembly: AssemblyDescription("Dot Net API for OWASP ZAP")]
 [assembly: AssemblyConfiguration("")]
-[assembly: AssemblyCompany("OWASP ZAP")]
-[assembly: AssemblyProduct("OWASPZAPDotNetAPI")]
-[assembly: AssemblyCopyright("Copyright ©  2015")]
+[assembly: AssemblyCompany("Monster Worldwide")]
```
Comment (thc202) [inline src/OWASPZAPDotNetAPI/OWASPZAPDotNetAPI/Properties/AssemblyInfo.cs]: Was this change on purpose?

File: README [hunk 4]
```diff
@@ -0,0 +1,2 @@
+.NET Core library for interacting with OWASP ZAP proxy.
+Fixes issue with api key and exposes api methods in async way.
```
Comment (thc202) [inline README]: Probably better to move this to a change log file?

**Comment by gmaran23** [comment]
@svejdo1 The removeContext method needs to be added to the API. And the new dll needs to be released in nuget.org.  I willrelease the recent dll as a nuget package.

However, the changes for async and await, I think can wait. One of the original rationale when writing the Dot Net API was to keep it simple and allow it to be consumed by people that have little or no experience about the new C# language features r the .Net framework new features. For example, LINQs weren't used at places, where we could have used it in ample, and then Task Parallel Library wasn't used where we could have used it, FluentAPI weren't used in the sample code.

The Dot Net API as it is (with very simple C# syntax, the most basic .Net base classes), can just be used by any consumer that does not need to have intermediate skills on Dot Net programming.

**Comment by svejdo1** [comment]
@gmaran23  - well you're out of luck in newer version of .NET - for example in .NET core you can't craft HttpRequest in a sync way.  You can hide away from it by using something like Task.Run()... in your library - but that's a bad policy in library class. And I think since OWASP itself uses java (multi-platform thingy) having a version of client library on .NET core (which is also multi-platform thingy) can be beneficial - heck - I'm running this library as part of some script in docker/ubuntu.

**Comment by gmaran23** [comment]
@svejdo1 ok. Makes sense now. However, I see some challenges, and also the advantage of using .Net Core (to run on multiple platforms).

If we just release a OWASPZAPDOTNETAPI.dll targeting .Net Core, then all the existing consumers that wants to consume new API features (like removeContext for example) need to upgrade their automation code with async await. Not very good idea with backward compatibility. It is a breaking change.

I propose maintaining two different dlls for now. OWASPZAPDOTNETAPI.dll as it is right now. OWASPZAPDOTNETAPIFORDOTNETCORE.dll (the name can obviously improve) specifically targeting the .Net Core. 

1. People that want to upgrade their automation just to take advantage of the recent API features could continue using OWASPZAPDOTNETAPI.dll. 

2. People that begin new automation could already start with OWASPZAPDOTNETAPIFORDOTNETCORE.dll. 

Having said that, it does require intermediate level .Net programming skills to use the async & await model. Most penetration testers with scripting experience are not comfortable with it. So, at least for a couple of years from now, both these dlls can co-exist. Maintaining two versions would not be a problem because the API files are autogenerated.

Any other thoughts?

**Comment by svejdo1** [comment]
Makes sense.

**Comment by thc202** [comment]
Closing, latest code now has the `removeContext`.

@svejdo1 if there are still changes in this pull request that you would like to apply, please, open and rebase, pull request now has conflicts (or, just open other pull request, whatever you prefer). Thanks!
